src/Microsoft.ComponentDetection.Detectors/dotnet/DotNetComponentDetector.cs

```diff
@@ -0,0 +1,186 @@
+    {
+        var lockFile = this.lockFileFormat.Read(processRequest.ComponentStream.Stream, processRequest.ComponentStream.Location);
+
+        var projectPath = lockFile.PackageSpec.RestoreMetadata.ProjectPath;
+        var projectDirectory = this.pathUtilityService.GetParentDirectory(projectPath);
+        var sdkVersion = await this.GetSdkVersionAsync(projectDirectory, cancellationToken);
+
+        var projectName = lockFile.PackageSpec.RestoreMetadata.ProjectName;
+        var projectOutputPath = lockFile.PackageSpec.RestoreMetadata.OutputPath;
+        var targetType = this.GetProjectType(projectOutputPath, projectName, cancellationToken);
+
+        var componentReporter = this.ComponentRecorder.CreateSingleFileComponentRecorder(projectPath);
+        foreach (var target in lockFile.Targets)
+        {
+            var targetFramework = target.TargetFramework?.GetShortFolderName();
+
+            componentReporter.RegisterUsage(new DetectedComponent(new DotNetComponent(sdkVersion, targetFramework, targetType)));
+        }
+    }
+
+    private string? GetProjectType(string projectOutputPath, string projectName, CancellationToken cancellationToken)
+    {
+        if (this.directoryUtilityService.Exists(projectOutputPath))
+        {
+            var namePattern = (projectName ?? "*") + ".dll";
+
+            // look for the compiled output, first as dll then as exe.
+            var candidates = this.directoryUtilityService.EnumerateFiles(projectOutputPath, namePattern, SearchOption.AllDirectories)
+                     .Concat(this.directoryUtilityService.EnumerateFiles(projectOutputPath, namePattern, SearchOption.AllDirectories));
+            foreach (var candidate in candidates)
+            {
+                if (this.IsApplication(candidate))
+                {
+                    return "application";
+                }
+                else
+                {
+                    return "library";
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private bool IsApplication(string assemblyPath)
+    {
+        try
+        {
+            using var peReader = new PEReader(this.fileUtilityService.MakeFileStream(assemblyPath));
+
+            // despite the name `IsExe` this is actually based of the CoffHeader Characteristics
+            return peReader.PEHeaders.IsExe;
+        }
+        catch (Exception)
+        {
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Recursively get the sdk version from the project directory or parent directories.
+    /// </summary>
+    /// <param name="projectDirectory">Directory to start the search.</param>
+    /// <param name="cancellationToken">Cancellation token to halt the search.</param>
+    /// <returns>Sdk version found, or null if no version can be detected.</returns>
+    private async Task<string?> GetSdkVersionAsync(string projectDirectory, CancellationToken cancellationToken)
```

Comment:
While this shouldn't happen, we should return null if projectDirectory is null or whitespace.